### ISSUE TITLE
Help on None type check

### ISSUE THREAD

**Issue by tsunamilx** [microsoft/pyright#10247]
Hi guys,

I am new to this, I usually dont use type check as I am a sort of casual developer, but recently I've encountered issues because I accidentally pass the wrong parameter to a function, that wrong parameter has different type, and IDE PyCharm is supposed to warn but it didn't. So I think it's the good time to use a more robust type check tool, so I am here.

There is one thing about the None type check that I dont like and I am asking if there is way to ignore it

```python
#  the following reports error: expect SomeType got None
var: SomeType = None

# have to write this:
var: SomeType | None = None

# when define a parameter: Type "None" is not assignable to type "SomeType"
def func(param: SomeType = None)

# have to write this:
def func(param: SomeType | None = None)

# when doing something with the `var`: Operator "+" not supported for "None"
var += other

# have to write this:
assert var is not None
var += other
# or
if var is not None:
    var += other
```

in the case of `var: SomeType = None`, for me it is clear that the var is optional or may be initialized later
in the case of `var += other`, for me it is nice to check None but I've made sure the var must not be None somewhere else, I dont want to check it every time I use it.

I know I can make specific rule ignored in the config file, but by doing so I am afraid that it may also silence when checking with non-None type, is there a way to ignore the cases I mentioned above with only None type?

Thanks.

**Comment by erictraut** [comment]
> is there a way to ignore the cases I mentioned above with only None type?

The short answer is no. If you declare that a variable can be assigned only `SomeType`, then it's the job (arguably the _core_ job) of a type checker to tell you when you violate this constraint. `None` is not assignable to this type. If you want a variable to accept either `SomeType` or `None`, then you must declare it as such.

For more information about type declarations and how they work, you may find [this documentation](https://microsoft.github.io/pyright/#/type-concepts) helpful. For a more in-depth understanding of assignability rules in the Python type system, you may want to read [this section of the Python typing spec](https://typing.python.org/en/latest/spec/concepts.html).

If a type checker determines that a variable can potentially contain `None` at a particular point in the code flow graph of a function, it will tell you if you're using it in a manner where `None` can result in a crash. If your code assumes that the value is not `None` at that point, you can either add a type guard condition (`if var is not None`) or an assert (`assert var is not None`) to document and enforce that assumption.

**Comment by tsunamilx** [comment]
OK, thanks for the explanation
